### PR TITLE
Add support for storing drupal vm in a subdirectory

### PR DESCRIPTION
**Pull request by oxyc**
This is a work in progress and currently (probably) doesn't work with:
- `ansible_local`
- correct `post_provisioning_scripts` paths

I'm adding my initial proof of concept described in https://github.com/geerlingguy/drupal-vm/issues/305#issuecomment-156702888 as well as some docs on how to set it up.

Would be great to get some feedback and suggestions, so I'm opening this PR for more discussion. Personally I really like this setup as it's fairly simple, doesn't add too much code and allows me to store Drupal VM as a git submodule.

In the discussion I linked to @frob discovered one drawback though. You need to provision the machine in the project root, as the `.vagrant` directory needs to be created in the same location as the delegating `Vagrantfile` or else multiple `.vagrant` folders will be created. Actually now that I think about it, it's probably only an issue if you go through the delegat**ed** Vagrantfile from Drupal VM.

I honestly don't think that scenario is worth fixing... But it could probably be prevented by doing something similar to http://stackoverflow.com/a/25904132/319855

### CODE REVIEW

**Comment by geerlingguy** [comment]
Hmm... this is an interesting approach, for sure, and would probably jive pretty well with the approach we usually use on Acquia's projects. I'm going to kick the tires on it, would also love to hear more feedback from others from #305 as well.

**Comment by oxyc** [comment]
Due to https://github.com/geerlingguy/drupal-vm/issues/372 I can't test this at the moment but I think 06aa960 would make `ansible_local` work.

**Comment by jyrkij** [comment]
Regarding config file path I'd like to be able to set an arbitrary path for that. Just to minimize the number of files in project root...

**Comment by frob** [comment]
I have not had a chance to fully test this, however I had a few thoughts. 

Could we use a symlink to the drupal-vm directory? I am still not comfortable with making anything other than the drupal-vm config a part of the actual project repository.

**Comment by oxyc** [comment]
@frob you mean symlinking the drupal-vm directory from outside the project root? I just tried it with rsynced folders and I only got the symlink to sync, but not the contents. I'm guessing this is something that can differ per provider and per synced folder type. Is this supposed to be supported? I don't really feel it's a blocker personally..

**Comment by oxyc** [comment]
I'll wait until I get `ansible_local` working again but I think `ENV` variables is probably a better solution. That way the wrapper doesn't even need to be a Vagrantfile, it could be a bash script or a gulp task or whatever.

**Comment by oxyc** [comment]
Added a commit that uses environment variables instead. Not sure what's going on with the `ansible_local` provisioner though, it might be an upstream bug because it fails while running the following command:
```sh
cd /vagrant && PYTHONUNBUFFERED=1 ANSIBLE_FORCE_COLOR=true ansible-playbook --limit='drupalvm' --inventory-file=/tmp/vagrant-ansible/inventory --extra-vars={"config_file":"/vagrant/config/config.yml"} box/provisioning/playbook.yml
```
Note that the --extra-vars value is not quoted and therefore throws an error. If I ssh in and modify the command to the following it works.
```sh
cd /vagrant && PYTHONUNBUFFERED=1 ANSIBLE_FORCE_COLOR=true ansible-playbook --limit='drupalvm' --inventory-file=/tmp/vagrant-ansible/inventory --extra-vars='{"config_file":"/vagrant/config/config.yml"}' box/provisioning/playbook.yml
```

Disregarding the issue I mentioned it works in the following scenarios:
1. Current setup (both `ansible` and `ansible_local` provisioners)
3. Drupal VM in a sub directory and config.yml in another subdirecotry (both `ansible` and `ansible_local` provisioners)

The paths are really messy though, I cant even follow it myself. I'm just pushing the proof of concept.

**Comment by frob** [comment]
I was reffering to symlinking to the .vagrant directory in the drupalvm directory from the project directory. This would be to stop the confusing vagrantfile proxy issue.

**Comment by oxyc** [comment]
Cleaned up a bit and found the upstream issue about `--extra-vars` not being quoted https://github.com/mitchellh/vagrant/issues/6726

**Comment by oxyc** [comment]
@frob symlinking works with VirtualBox at least.

```sh
.
├── .vagrant
│   ├── machines
│   └── provisioners
├── Vagrantfile # Delegates to box/Vagrantfile
├── box # Drupal VM checkout
│   ├── .git
│   ├── .gitignore
│   ├── .travis.yml
│   ├── .vagrant -> ../.vagrant
│   ├── LICENSE
│   ├── README.md
│   ├── Vagrantfile
│   ├── config.yml -> ../config/config.yml
│   ├── docs
│   ├── example.config.yml
│   ├── example.drupal.make.yml
│   ├── examples
│   ├── mkdocs.yml
│   └── provisioning
└── config
    └── config.yml
```

Steps:
```sh
pushd box;
ln -s ../.vagrant .vagrant
ln -s ../config/config.yml .
popd
```

It works as long as you provision it in the project root the first time. Guess it works in the opposite direction too.

Currently I'm quite happy with this setup. Only waiting for the vagrant issue to be solved. Hopefully it gets fixed by 1.8.2

**Comment by oxyc** [comment]
@frob symlinking works with VirtualBox on OSX at least (I do not have access to any other setup).

```sh
.
├── .vagrant
│   ├── machines
│   └── provisioners
├── Vagrantfile # Delegates to box/Vagrantfile
├── box # Drupal VM checkout
│   ├── .git
│   ├── .gitignore
│   ├── .travis.yml
│   ├── .vagrant -> ../.vagrant
│   ├── LICENSE
│   ├── README.md
│   ├── Vagrantfile
│   ├── config.yml -> ../config/config.yml
│   ├── docs
│   ├── example.config.yml
│   ├── example.drupal.make.yml
│   ├── examples
│   ├── mkdocs.yml
│   └── provisioning
└── config
    └── config.yml
```

Steps:
```sh
pushd box
ln -s ../.vagrant .vagrant
ln -s ../config/config.yml config.yml
```

It works as long as you provision it in the project root the first time. Guess it works in the opposite direction too. BUT the config file now has to be inside the drupal vm directory again! No way to go around this (some more advanced setup could set the `ENV` variables conditionally I guess) as the path is no longer available from the delegating `Vagrantfile`. I think this is a bit unnecessary and we could just mention in the docs that the user shouldn't run commands through `box/Vagrantfile`.

Currently I'm quite pleased with this setup. Only waiting for the vagrant issue to be solved. Hopefully it gets fixed by 1.8.2

**Comment by oxyc** [comment]
@geerlingguy and anyone else, please give further input and I can make adjustments if necessary. Before merging I'll squash the commits but I'll keep them separated for now so you can see the changes I've made.

**Comment by oxyc** [comment]
Postponed by https://github.com/geerlingguy/drupal-vm/issues/450

**Comment by oxyc** [comment]
I still consider this postponed until Vagrant `1.8.2` (the `--extra-vars` bug was added to the milestone).
Just thinking about the patch in general. Maybe it's not worth adding support for a custom playbook. Sounds like you should be maintaining a fork if that's something which is interesting.

**Comment by oxyc** [comment]
One way to avoid having to define `DRUPALVM_PROJECT_ROOT`, would be to upload the config file using Vagrants `file` provisioner when using `ansible_local`. This would be the entire change of the Vagrantfile (compared to what this current PR is).

```diff
diff --git a/Vagrantfile b/Vagrantfile
index b872b14..94f0d00 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -27,10 +27,11 @@ end
 # Use config.yml for basic VM configuration.
 require 'yaml'
 dir = File.dirname(File.expand_path(__FILE__))
-unless File.exist?("#{dir}/config.yml")
+config_file = ENV['DRUPALVM_CONFIG'] || "#{dir}/config.yml"
+unless File.exist?(config_file)
   raise 'Configuration file not found! Please copy example.config.yml to config.yml and try again.'
 end
-vconfig = YAML.load_file("#{dir}/config.yml")
+vconfig = YAML.load_file(config_file)

 # Replace jinja variables in config.
 vconfig = walk(vconfig) do |value|
@@ -112,6 +113,9 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   if which('ansible-playbook')
     config.vm.provision 'ansible' do |ansible|
       ansible.playbook = "#{dir}/provisioning/playbook.yml"
+      ansible.extra_vars = {
+        "config_file": config_file
+      }
     end
   else
     config.vm.provision 'shell' do |sh|
@@ -121,9 +125,13 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
   # ansible_local provisioner is broken in Vagrant < 1.8.2.
   # else
+  #   config.vm.provision "file", souce: config_file, destination: "/tmp/config.yml"
   #   config.vm.provision "ansible_local" do |ansible|
   #     ansible.playbook = "provisioning/playbook.yml"
   #     ansible.galaxy_role_file = "provisioning/requirements.yml"
+  #     ansible.extra_vars = {
+  #       "config_file": "/tmp/config.yml"
+  #     }
   #   end
   # end
```

You would use a delegating Vagrantfile with:
```ruby
ENV['DRUPALVM_CONFIG'] = "#{__dir__}/config/drupal-vm.config.yml"
load "box/Vagrantfile"
```

**Comment by oxyc** [comment]
Here's a far simpler patch which only tackles the issue of storing Drupal VM in a subdirectory.

This should work already as we aren't relying on `ansible_local` anymore. Currently it fails due to the upstream Galaxy issue though. I'll test it and make sure it works as soon as the galaxy bug is fixed and I have some spare time again.

@geerlingguy what are your thoughts about this? By only adding support for a custom config location this patch is a lot smaller and less scary. 

https://github.com/geerlingguy/drupal-vm/compare/master...oxyc:subdir

**Comment by geerlingguy** [comment]
@oxyc - I do like having it a bit simpler, though an environment variable could be slightly perplexing if you have multiple Drupal VM instances in different subdirs...

Also, any changes to `provisioning/JJG-Ansible-Windows/windows.sh` should be submitted against the upstream project [JJG-Ansible-Windows](https://github.com/geerlingguy/JJG-Ansible-Windows).

**Comment by oxyc** [comment]
> Also, any changes to provisioning/JJG-Ansible-Windows/windows.sh should be submitted against the upstream project JJG-Ansible-Windows.

Ah I wasn't aware, if you're interesting in getting this in before we move back to `ansible_local` I can clean it up as a generalized `extra_vars` flag.

The environment variable isn't exported though so it's just available within the process. Same as the general unix convention of `DEBUG=1 cmd`. This would also work:
`DRUPALVM_CONFIG="~/foo/config.yml" vagrant up`.

**Comment by oxyc** [comment]
So back to having to define `DRUPALVM_PROJECT_ROOT`. It really is required to know the location of the parent Vagrantfile and I can't find a way a way to without manually specifying it:

```ruby
ENV['DRUPALVM_PROJECT_ROOT'] = "#{__dir__}"
ENV['DRUPALVM_CONFIG'] = "#{__dir__}/config/drupal-vm.config.yml"
load "box/Vagrantfile"
```

I feel like I've exhausted all my ideas, there doesn't seem to be an easier way to implement it really. Happy to rework the PR if someone has some feedback though.

**Comment by joestewart** [comment]
@oxyc This is looking great.  Worked after some small changes but have only lightly tested.  Will be nice to have a local.config.yml with the box in a subdir too.  thanks.

I had to make these changes:

```
diff --git a/Vagrantfile b/Vagrantfile
index 9fef925..2ada931 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -20,7 +20,7 @@ host_project_root   = ENV['DRUPALVM_PROJECT_ROOT'] || host_drupalvm_root
 guest_drupalvm_root = "/vagrant/#{Pathname.new(host_drupalvm_root).relative_path_from(Pathname.new(host_project_root))}"
 guest_project_root  = "/vagrant"
 
-host_vconfig_path       = ENV['DRUPALVM_CONFIG']       ? "#{host_project_root}/#{ENV['DRUPALVM_CONFIG']}"        : "#{host_drupalvm_root}/config.yml"
+host_vconfig_path       = ENV['DRUPALVM_CONFIG']       ? "#{ENV['DRUPALVM_CONFIG']}"        : "#{host_drupalvm_root}/config.yml"
 host_playbook_path      = ENV['DRUPALVM_PLAYBOOK']     ? "#{host_project_root}/#{ENV['DRUPALVM_PLAYBOOK']}"      : "#{host_drupalvm_root}/provisioning/playbook.yml"
 guest_vconfig_path      = ENV['DRUPALVM_CONFIG']       ? "#{guest_project_root}/#{ENV['DRUPALVM_CONFIG']}"       : "#{guest_drupalvm_root}/config.yml"
 guest_playbook_path     = ENV['DRUPALVM_PLAYBOOK']     ? "#{guest_project_root}/#{ENV['DRUPALVM_PLAYBOOK']}"     : "#{guest_drupalvm_root}/provisioning/playbook.yml"
@@ -28,6 +28,7 @@ guest_requirements_path = ENV['DRUPALVM_REQUIREMENTS'] ? "#{guest_project_root}/
 
 # Use config.yml for basic VM configuration.
 require 'yaml'
+print
 if !File.exist?(host_vconfig_path)
   raise 'Configuration file not found! Please copy example.config.yml to config.yml and try again.'
 end
@@ -100,7 +101,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     config.vm.provision "ansible" do |ansible|
       ansible.playbook = host_playbook_path
       ansible.extra_vars = {
-        "config_file": host_vconfig_path
+        "config_file" => host_vconfig_path
       }
     end
   else
@@ -108,7 +109,7 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
       ansible.playbook = guest_playbook_path
       ansible.galaxy_role_file = guest_requirements_path
       ansible.extra_vars = {
-        "config_file": guest_vconfig_path
+        "config_file" => guest_vconfig_path
       }
     end
   end
```

File: provisioning/JJG-Ansible-Windows/windows.sh [hunk 1]
```diff
@@ -17,7 +34,7 @@ YUM=$(which yum 2>/dev/null)
 APT_GET=$(which apt-get 2>/dev/null)
 
 # Make sure Ansible playbook exists.
-if [ ! -f "/vagrant/$ANSIBLE_PLAYBOOK" ]; then
```
Comment (oxyc) [inline provisioning/JJG-Ansible-Windows/windows.sh]: If this is acceptable, we need to commit this to https://github.com/geerlingguy/JJG-Ansible-Windows first. I think forcing  the path to be within `/vagrant` to be an unnecessary restriction. Also it looks weird when you call the script with an absolute-looking path that's actually relative.

File: Vagrantfile [hunk 2]
```diff
@@ -103,31 +111,37 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   end
 
   # Allow override of the default synced folder type.
-  config.vm.synced_folder '.', '/vagrant', type: vconfig.include?('vagrant_synced_folder_default_type') ? vconfig['vagrant_synced_folder_default_type'] : 'nfs'
+  config.vm.synced_folder host_project_root, '/vagrant', type: vconfig.include?('vagrant_synced_folder_default_type') ? vconfig['vagrant_synced_folder_default_type'] : 'nfs'
```
Comment (oxyc) [inline Vagrantfile]: Now it supports provisioning anywhere in the project tree.

**Comment by oxyc** [comment]
@joestewart thanks for testing! I've rebased and cleaned everything up so it's easier for everyone to test.

I've settled on `DRUPALVM_CONFIG` path being relative, so you need to change that. We could always add a statement that checks if it's absolute and makes it relative to support both.

I've tested:
- Default usage without a delegating Vagrantfile, both ansible and shell provisioner
- New usage with a delegating Vagrantfile, both ansible and shell provisioner

File: provisioning/JJG-Ansible-Windows/windows.sh (hunk 1 above)
Comment (geerlingguy) [inline provisioning/JJG-Ansible-Windows/windows.sh]: For sure - and I'm definitely open to that.

**Comment by geerlingguy** [comment]
This is one of the top priority items I hope to tackle before/during DrupalCon... I'll be on vacation next week, and all around the US the next couple weeks, so I probably won't be diving too deep into the changes/review until just before DC.

**Comment by oxyc** [comment]
I was considering how to add support for a modified `requirements.yml` file https://github.com/geerlingguy/drupal-vm/issues/505, and only way I can think of is to once again check for a `local.requirements.yml` file. The most annoying part about this patch is the logic required for storing the `config.yml` in an arbitrary subdirectory. If we add support for `local.requirements.yml` we would need to do the same path checks again.

Forcing the the files to be in a set location, or _at least_ in the same location would make life easier. Eg.
- same directory as the `Vagrantfile`
- a `.drupalvm/` directory.
- a user definable config directory, but shared for all configs.

Worth considering before continuing with this PR.

**Comment by oxyc** [comment]
I'm testing out an approach with specifying a `DRUPALVM_CONFIG_DIR` instead of individual config file paths.

Here's what I've come up with:

1. A nth version of this PR https://github.com/oxyc/drupal-vm/commit/8e098c9248bbf786111f6c6f1381980d08c59db8.

2. I modified our current base repo, to use Drupal VM as a composer dependency and removed my old hacks with the makefile (PR fixes that) https://github.com/generoi/drupal-project/commit/ab86474fea004d4042a31b7c4bece89e9386c21e

3. Added a `local.config.yml` for setting up a test VM https://github.com/generoi/drupal-project/commit/2fca346038c1770f3735a30de0a50aca25f20006 (I'm liking local.config.yml a lot!)

I briefly tested my fork and made sure it worked with the new setup, with the default setup and with/without JJG-Ansible-Windows. I'm planning to provision and install the entire drupal vm for each scenario during the day though.

**Comment by oxyc** [comment]
@geerlingguy I would love your input on this PR. What are your concerns / what would be needed to get it merged in 3.0?

**Comment by oxyc** [comment]
Docs based on config directory idea + composer dependency setup: https://gist.github.com/oxyc/f932c551b1e17c977da2db891984a38d

**Comment by geerlingguy** [comment]
I will look at this in a couple minutes!

**Comment by geerlingguy** [comment]
+1, let's do it

File: provisioning/JJG-Ansible-Windows/windows.sh (hunk 1 above)
Comment (geerlingguy) [inline provisioning/JJG-Ansible-Windows/windows.sh]: Do we still need to make a change upstream?

Comment (oxyc) [inline provisioning/JJG-Ansible-Windows/windows.sh]: Ah yes we do. I'll submit a PR tonight. I have some other things to take care of right now but I'll have my own sprint in the evening.

Maybe we technically can avoid it but I feel it adds a lot more readability to what those variables are.

Comment (oxyc) [inline provisioning/JJG-Ansible-Windows/windows.sh]: https://github.com/geerlingguy/JJG-Ansible-Windows/pull/27

Comment (oxyc) [inline provisioning/JJG-Ansible-Windows/windows.sh]: If you prefer not merging that PR, it's just a matter of changing:

```rb
guest_drupalvm_dir = ENV['DRUPALVM_DIR'] ? "/vagrant/#{ENV['DRUPALVM_DIR']}" : guest_project_dir
```

to

```rb
guest_drupalvm_dir = ENV['DRUPALVM_DIR'] ? ENV['DRUPALVM_DIR'] : ''
```

And then we change it back once Vagrant 1.8.2 comes out.

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            "homepage": "http://www.jeffgeerling.com"
+        }
+    ],
+    "support": {
+        "issues": "https://github.com/geerlingguy/drupal-vm/issues",
+        "source": "https://github.com/geerlingguy/drupal-vm",
+        "docs": "http://docs.drupalvm.com"
+    },
+    "require": {}
+}
```
Comment (oxyc) [inline composer.json]: For reference here are two other VM packages published to Packagist:
https://github.com/laravel/homestead/blob/master/composer.json
https://github.com/roots/bedrock/blob/master/composer.json

**Comment by oxyc** [comment]
Just realized we can do `composer create-project geerlingguy/drupal-vm:2.5.1 drupalvm` as well right?

File: composer.json (hunk 3 above)
Comment (geerlingguy) [inline composer.json]: @oxyc - If you'd like to add yourself as another author (for the credit), that's fine by me!

**Comment by geerlingguy** [comment]
@oxyc - Yeah, that should also work, for quick Drupal project setup, or things like "Build me a quick Drupal environment for core hacking".

Quick start guide could become a lot shorter :)

**Comment by geerlingguy** [comment]
Of course... one thing about the current quick start guide—you don't need composer or PHP installed on your host at all to use Drupal VM or start developing.

So we could offer that as a "super duper quick start for those with `composer`" or something.

File: composer.json (hunk 3 above)
Comment (oxyc) [inline composer.json]: Thanks. Sure, some free inbound links once I eventually work a development blog/site hehe.

Comment (geerlingguy) [inline composer.json]: Ha! You should start blogging—just posting some of the cool stuff you guys do for local dev would be worth some inbound links!

**Comment by geerlingguy** [comment]
I'll either review this in a few minutes (after I grab a snack), or tomorrow if said snack results in my falling asleep while watching some TV :P

**Comment by geerlingguy** [comment]
@oxyc - D'oh, I think I overwrote the JJG-Ansible-Windows changes you required since I updated the upstream project.

**Comment by oxyc** [comment]
Rebased and my version of that commit disappeared. So good to go if no other changes are needed.

**Comment by oxyc** [comment]
Or no. Found an issue.

**Comment by geerlingguy** [comment]
K, I'm fading fast, so I'll take another look tomorrow. Thanks!

**Comment by oxyc** [comment]
Yeah, this is a pretty big change so better to review it properly.

Btw. for quickly testing different setups you can do stuff like:
```
mkdir config-scenario
mkdir config-second-scenario
cp example.config.yml config-scenario/
cp example.config.yml config-second-scenario/

DRUPALVM_PROJECT_ROOT="$(pwd)" DRUPALVM_DIR="." DRUPALVM_CONFIG_DIR="config-scenario" vagrant provision
```
Pretty handy to be honest.

**Comment by oxyc** [comment]
To make `example.config.yml` work out of the box in a subdirectory we can change `drush_makefile_path` to

```yaml
drush_makefile_path: "{{ config_dir }}/drupal.make.yml"
```

Then it assumes it's located in the same directory as `config.yml`.

**Comment by oxyc** [comment]
I updated the `drush_makefile_path` and the docs but if you prefer I can revert that commit.

**Comment by geerlingguy** [comment]
@oxyc - You know, one other thing that I'm wondering... with our new `with_fileglob` setup, why not make `config.yml` optional as well. Load in `example.config.yml` (maybe name it `default.config.yml`) as the base, since it will always be present, then load config.yml, local.config.yml (and in the future others) iff they exist? (Probably something for a different issue, but I'm still reading through here right now).

**Comment by oxyc** [comment]
I was considering using `config.yml` as a symlink to Drupal VMs `example.config.yml` and use `local.config.yml` for our tem config so that people at work have easier keeping track of what's different and hopefully so they can do the upgrades themselves. Your idea would work very nicely for this setup.

That's also something I liked about beetbox.

**Comment by geerlingguy** [comment]
@oxyc I'm going to merge this as-is—can you maybe open a follow-up issue or PR to start using `example.config.yml` as the default (so `config.yml` is optional)?

**Comment by geerlingguy** [comment]
Now I start the gauntlet of tests on my Mac and Windows laptops... please everyone test in different scenarios to see what might be broken, so we cut a nice clean 3.0.0 release!

**Comment by oxyc** [comment]
> @oxyc I'm going to merge this as-is—can you maybe open a follow-up issue or PR to start using example.config.yml as the default (so config.yml is optional)?

Sure, I'll work on getting a PR sent over a bit later today.

**Comment by oxyc** [comment]
Also yey! This is the following up on my first PR #147  😃

**Comment by oxyc** [comment]
I tested and got one of our client projects set up using Drupal VM as a composer dependency by following the docs. Awesome!

**Comment by geerlingguy** [comment]
Awesome! I'll try to tag 3.0.0 in a little bit. Need to get my notes together first...

**Pull request merged**